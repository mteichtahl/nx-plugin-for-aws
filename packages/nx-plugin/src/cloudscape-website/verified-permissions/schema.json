{
  "$schema": "https://json-schema.org/schema",
  "$id": "Verified permissions",
  "cli": "nx",
  "title": "AWS Verified Permissions for Cloudscape website.",
  "description": "Add AWS Verified Permissions to an existing ts#cloudscape-website#cognitoAuth",
  "type": "object",
  "properties": {
    "project": {
      "description": "The root directory of the Cloudscape application.",
      "type": "string",
      "alias": "dir",
      "x-priority": "important",
      "x-prompt": "The root directory for your Cloudscape application",
      "x-dropdown": "projects"
    },
    "namespace": {
      "description": "Namespace of the verified permissions schema. Can not be aws, amazon, or cedar.",
      "type": "string",
      "x-prompt": "What is the namespace for your verified permissions?",
      "x-priority": "important",
      "default": "Default"
    },
    "principalEntity": {
      "description": "The principal entity type that you want to associate with your identity source.",
      "type": "string",
      "x-prompt": "What is your principal entity?",
      "x-priority": "important",
      "default": "User"
    },
    "groupEntity": {
      "description": "The group entity type that you want to associate with your groups.",
      "type": "string",
      "x-prompt": "What is your group entity?",
      "x-priority": "important",
      "default": "Group"
    }
  },
  "required": ["project"]
}
