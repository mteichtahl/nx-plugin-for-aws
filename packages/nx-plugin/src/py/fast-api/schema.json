{
  "$schema": "https://json-schema.org/schema",
  "$id": "PyFastApi",
  "title": "Create a Python FastAPI project",
  "description": "Create a Python FastAPI project",
  "type": "object",
  "properties": {
    "name": {
      "type": "string",
      "description": "Name of the API project to generate",
      "x-priority": "important",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "What name would you like your API to have? i.e: MyApi"
    },
    "computeType": {
      "type": "string",
      "description": "The type of compute to use to deploy this API. Choose between ServerlessApiGatewayRestApi (default) or ServerlessApiGatewayHttpApi.",
      "default": "ServerlessApiGatewayRestApi",
      "enum": ["ServerlessApiGatewayRestApi", "ServerlessApiGatewayHttpApi"],
      "x-prompt": "What compute type would you like to deploy your API with?",
      "x-priority": "important"
    },
    "auth": {
      "type": "string",
      "description": "The method used to authenticate with your API. Choose between IAM (default), Cognito or None.",
      "default": "IAM",
      "enum": ["IAM", "Cognito", "None"],
      "x-prompt": "How would you like users to authenticate with your API?",
      "x-priority": "important"
    },
    "directory": {
      "type": "string",
      "description": "The directory to store the application in.",
      "default": "packages",
      "x-prompt": "Which directory do you want to create the project in?",
      "x-priority": "important"
    },
    "moduleName": {
      "type": "string",
      "description": "Python module name"
    }
  },
  "required": ["name"]
}
