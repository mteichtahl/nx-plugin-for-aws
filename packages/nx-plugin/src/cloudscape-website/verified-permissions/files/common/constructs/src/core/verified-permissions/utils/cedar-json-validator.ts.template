/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
/**
 * Valid scalar attribute types supported by Cedar
 */
export type ScalarAttributeType = 'String' | 'Long' | 'Boolean' | 'Double';

/**
 * Valid collection attribute types supported by Cedar
 */
export type CollectionAttributeType = 'Set' | 'Record';

/**
 * All valid attribute types supported by Cedar
 */
export type ValidAttributeType = ScalarAttributeType | CollectionAttributeType;

/**
 * Element definition for collection types
 */
export interface AttributeElement {
  /** The data type of elements in the collection */
  type: string;
  [key: string]: unknown;
}

/**
 * Defines the structure of an attribute in a Cedar schema.
 * Attributes can be scalar types like String or Long, or collections like Sets.
 */
export interface AttributeType {
  /** The data type of this attribute (e.g., "String", "Long", "Set") */
  type: string;

  /** Whether this attribute is required */
  required?: boolean;

  /** For collection types like Set, defines the type of elements in the collection */
  element?: AttributeElement;

  [key: string]: unknown;
}

/**
 * Shape definition for entity types
 */
export interface EntityShape {
  /** The type of shape (usually "Record" in Cedar schemas) */
  type: string;

  /** Attributes of this entity type */
  attributes?: Record<string, AttributeType>;
}

/**
 * Defines the structure of an entity type in the Cedar schema.
 * Entity types can represent principals (like Users) or resources (like Documents).
 */
export interface EntityTypeDefinition {
  /** Types that this entity can be a member of (for group membership) */
  memberOfTypes?: string[];

  /** The structure of this entity type, including its attributes */
  shape?: EntityShape;
}

/**
 * Defines the structure of an action in the Cedar schema.
 * Actions represent operations that principals can perform on resources.
 */
export interface ActionDefinition {
  /** Specifies which principal types can perform this action on which resource types */
  appliesTo?: {
    /** Resource types that this action can be performed on */
    resourceTypes?: string[];

    /** Principal types that can perform this action */
    principalTypes?: string[];

    /** Context attributes that may be relevant for this action */
    context?: Record<string, AttributeType>;
  };
}

/**
 * Represents the complete structure of a Cedar JSON schema for authorization modeling.
 * This schema defines entity types, their shapes, attributes, and relationships,
 * as well as actions that can be performed within the authorization model.
 */
export interface CedarJson {
  /** Entity types in the Cedar schema, defining principals and resources */
  entityTypes?: Record<string, EntityTypeDefinition>;

  /** Actions that can be performed by principals on resources */
  actions?: Record<string, ActionDefinition>;

  /** Allow for other top-level properties that might be needed */
  [key: string]: unknown;
}

/**
 * Validates a Cedar JSON schema against the expected format and requirements.
 * Performs comprehensive validation of entity types, their attributes, relationships,
 * and the actions that can be performed in the authorization model.
 *
 * @param schema - The Cedar schema in JSON format to validate
 * @throws Error if the schema does not match the required format
 */
export function validateCedarSchema(schema: CedarJson): void {
  if (!schema) {
    throw new Error('Cedar schema cannot be null or undefined');
  }

  // Check if schema is an object
  if (typeof schema !== 'object' || Array.isArray(schema)) {
    throw new Error('Cedar schema must be a valid JSON object');
  }

  // Validate entityTypes if present
  if (schema.entityTypes) {
    validateEntityTypes(schema.entityTypes);

    // Validate cross-references between entity types (like memberOfTypes)
    validateEntityRelationships(schema.entityTypes);
  }

  // Validate actions if present
  if (schema.actions) {
    validateActions(schema.actions);

    // Validate cross-references between actions and entity types
    if (schema.entityTypes) {
      validateActionEntityReferences(schema.actions, schema.entityTypes);
    }
  }
}

/**
 * Validates entity types defined in the Cedar schema.
 *
 * @param entityTypes - The entity types to validate
 * @throws Error if entity types don't meet the requirements
 */
function validateEntityTypes(entityTypes: Record<string, EntityTypeDefinition>): void {
  if (typeof entityTypes !== 'object' || Array.isArray(entityTypes)) {
    throw new Error('entityTypes must be a valid JSON object');
  }

  for (const [name, definition] of Object.entries(entityTypes)) {
    if (!name || name.trim() === '') {
      throw new Error('Entity type name cannot be empty');
    }

    // Ensure only valid properties are defined
    const validProperties = ['memberOfTypes', 'shape'];
    const definitionKeys = Object.keys(definition);

    for (const key of definitionKeys) {
      if (!validProperties.includes(key)) {
        throw new Error(`Invalid property '${key}' found in entity type '${name}'. Only 'memberOfTypes' and 'shape' are allowed.`);
      }
    }

    // Validate memberOfTypes if present
    if (definition.memberOfTypes) {
      if (!Array.isArray(definition.memberOfTypes)) {
        throw new Error(`memberOfTypes for entity ${name} must be an array`);
      }
      // Validate that each memberOfType is a string and not empty
      for (const memberType of definition.memberOfTypes) {
        if (typeof memberType !== 'string' || memberType.trim() === '') {
          throw new Error(`memberOfType for entity ${name} contains invalid or empty value`);
        }
      }
    }

    // Validate shape if present
    if (definition.shape) {
      if (typeof definition.shape !== 'object' || Array.isArray(definition.shape)) {
        throw new Error(`shape for entity ${name} must be a valid JSON object`);
      }

      // Validate that shape has a valid type
      if (!definition.shape.type || typeof definition.shape.type !== 'string') {
        throw new Error(`Shape for entity ${name} must have a valid type`);
      }

      // Validate that type is 'Record' (currently only Record types are supported in Cedar)
      if (definition.shape.type !== 'Record') {
        throw new Error(`Shape for entity ${name} must be of type 'Record', got '${definition.shape.type}'`);
      }

      // Validate attributes if present
      if (definition.shape.attributes) {
        if (typeof definition.shape.attributes !== 'object' || Array.isArray(definition.shape.attributes)) {
          throw new Error(`Attributes for entity ${name} must be a valid JSON object`);
        }

        // Validate each attribute's structure
        for (const [attrName, attrType] of Object.entries(definition.shape.attributes)) {
          if (!attrName || attrName.trim() === '') {
            throw new Error(`Entity ${name} has an attribute with empty name`);
          }

          if (!attrType.type || typeof attrType.type !== 'string') {
            throw new Error(`Attribute ${attrName} in entity ${name} must have a valid type`);
          }

          // Validate common attribute types
          validateAttributeType(attrName, attrType, name);
        }
      }
    }
  }
}

/**
 * Validates the type of an individual attribute.
 *
 * @param attrName - The name of the attribute
 * @param attrType - The attribute type definition
 * @param entityName - The name of the entity containing this attribute
 * @throws Error if the attribute type is invalid
 */
function validateAttributeType(attrName: string, attrType: AttributeType, entityName: string): void {
  // Validate based on the attribute type
  const validScalarTypes: ScalarAttributeType[] = ['String', 'Long', 'Boolean', 'Double'];
  const validCollectionTypes: CollectionAttributeType[] = ['Set', 'Record'];
  const validTypes = [...validScalarTypes, ...validCollectionTypes];

  if (!validTypes.includes(attrType.type as ValidAttributeType)) {
    throw new Error(
      `Attribute ${attrName} in entity ${entityName} has invalid type '${attrType.type}'. Valid types are: ${validTypes.join(', ')}`
    );
  }

  if (validScalarTypes.includes(attrType.type as ScalarAttributeType)) {
    // Scalar types don't need element validation
    if (attrType.element) {
      throw new Error(`Scalar attribute ${attrName} in entity ${entityName} should not have an element property`);
    }
  } else if (validCollectionTypes.includes(attrType.type as CollectionAttributeType)) {
    // Collection types require element validation
    if (attrType.type === 'Set') {
      if (!attrType.element) {
        throw new Error(`Set attribute ${attrName} in entity ${entityName} must have an element definition`);
      }

      if (typeof attrType.element !== 'object') {
        throw new Error(`Element definition for attribute ${attrName} in entity ${entityName} must be an object`);
      }

      if (!attrType.element.type || typeof attrType.element.type !== 'string') {
        throw new Error(`Element type for attribute ${attrName} in entity ${entityName} must be specified`);
      }

      // Validate that the element type is a valid scalar type
      if (!validScalarTypes.includes(attrType.element.type as ScalarAttributeType)) {
        throw new Error(
          `Element type for attribute ${attrName} in entity ${entityName} must be one of: ${validScalarTypes.join(', ')}, got '${attrType.element.type}'`
        );
      }
    }
  }
}

/**
 * Validates that entity relationships (like memberOfTypes references) are consistent.
 *
 * @param entityTypes - The entity types to validate relationships for
 * @throws Error if entity relationships are inconsistent
 */
function validateEntityRelationships(entityTypes: Record<string, EntityTypeDefinition>): void {
  const entityNames = Object.keys(entityTypes);
  for (const [entityName, definition] of Object.entries(entityTypes)) {
    // Check that all memberOfTypes references exist in the schema
    if (definition.memberOfTypes) {
      for (const memberType of definition.memberOfTypes) {
        if (!entityNames.includes(memberType)) {
          throw new Error(`Entity ${entityName} references non-existent entity type '${memberType}' in memberOfTypes`);
        }
      }
    }
  }
}

/**
 * Validates actions defined in the Cedar schema.
 *
 * @param actions - The actions to validate
 * @throws Error if actions don't meet the requirements
 */
function validateActions(actions: Record<string, ActionDefinition>): void {
  if (typeof actions !== 'object' || Array.isArray(actions)) {
    throw new Error('actions must be a valid JSON object');
  }

  for (const [name, definition] of Object.entries(actions)) {
    if (!name || name.trim() === '') {
      throw new Error('Action name cannot be empty');
    }

    // Validate appliesTo if present
    if (definition.appliesTo) {
      const { resourceTypes, principalTypes, context } = definition.appliesTo;

      // Validate resourceTypes if present
      if (resourceTypes !== undefined) {
        if (!Array.isArray(resourceTypes)) {
          throw new Error(`resourceTypes for action ${name} must be an array`);
        }
        // Check that resourceType entries are non-empty strings
        for (const resourceType of resourceTypes) {
          if (typeof resourceType !== 'string' || resourceType.trim() === '') {
            throw new Error(`resourceTypes for action ${name} contains invalid or empty value`);
          }
        }
      }

      // Validate principalTypes if present
      if (principalTypes !== undefined) {
        if (!Array.isArray(principalTypes)) {
          throw new Error(`principalTypes for action ${name} must be an array`);
        }
        // Check that principalType entries are non-empty strings
        for (const principalType of principalTypes) {
          if (typeof principalType !== 'string' || principalType.trim() === '') {
            throw new Error(`principalTypes for action ${name} contains invalid or empty value`);
          }
        }
      }

      // Validate context if present
      if (context !== undefined) {
        if (typeof context !== 'object' || Array.isArray(context)) {
          throw new Error(`context for action ${name} must be a valid JSON object`);
        }

        for (const [ctxName, ctxType] of Object.entries(context)) {
          if (!ctxName || ctxName.trim() === '') {
            throw new Error(`Action ${name} has a context attribute with empty name`);
          }

          if (!ctxType.type || typeof ctxType.type !== 'string') {
            throw new Error(`Context attribute ${ctxName} in action ${name} must have a valid type`);
          }
          // Validate the context attribute types
          validateAttributeType(ctxName, ctxType, `action ${name} context`);
        }
      }
    }
  }
}

/**
 * Validates that action entity type references are consistent with defined entity types.
 *
 * @param actions - The actions to validate
 * @param entityTypes - The entity types to check against
 * @throws Error if actions reference undefined entity types
 */
function validateActionEntityReferences(
  actions: Record<string, ActionDefinition>,
  entityTypes: Record<string, EntityTypeDefinition>
): void {
  const entityNames = Object.keys(entityTypes);
  for (const [actionName, definition] of Object.entries(actions)) {
    if (definition.appliesTo) {
      const { resourceTypes, principalTypes } = definition.appliesTo;
      // Check that all resourceTypes exist in the schema
      if (resourceTypes) {
        for (const resourceType of resourceTypes) {
          if (!entityNames.includes(resourceType)) {
            throw new Error(`Action ${actionName} references non-existent resource type '${resourceType}'`);
          }
        }
      }
      // Check that all principalTypes exist in the schema
      if (principalTypes) {
        for (const principalType of principalTypes) {
          if (!entityNames.includes(principalType)) {
            throw new Error(`Action ${actionName} references non-existent principal type '${principalType}'`);
          }
        }
      }
    }
  }
}
