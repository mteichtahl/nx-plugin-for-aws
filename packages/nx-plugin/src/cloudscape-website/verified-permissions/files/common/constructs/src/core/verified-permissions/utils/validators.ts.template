/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { ICustomAttribute, UserPool } from 'aws-cdk-lib/aws-cognito';

/**
 * Options for validating schema attributes against a Cognito User Pool
 */
export interface SchemaValidationOptions {
  /** The Cognito User Pool to check for attribute definitions */
  userPool: UserPool;
  /** The required attributes that should be defined in the user pool */
  requiredAttributes: {
    [key: string]: ICustomAttribute;
  };
}

/**
 * Validates that all required attributes are properly defined in the Cognito User Pool schema.
 *
 * This ensures that the attributes used in Cedar policies are actually available in Cognito,
 * preventing runtime errors when attributes are referenced but don't exist in the identity source.
 *
 * The validation checks:
 * - Presence of the attributes in the options
 * - Proper configuration of each attribute
 *
 * @param options - The validation options containing the user pool and required attributes
 * @throws Error if any required attribute is not properly defined in the user pool
 */
export function validateSchemaAttributes(options: SchemaValidationOptions): void {
  const { userPool, requiredAttributes } = options;

  if (!userPool) {
    throw new Error('User pool must be provided for attribute validation');
  }

  if (!requiredAttributes || Object.keys(requiredAttributes).length === 0) {
    return; // No required attributes to validate
  }

  // Extract the user pool's schema attributes for reference checking
  // Note: CDK doesn't expose the schema property directly, so we warn about attributes
  // that might be missing but can't validate them statically

  // Instead, we encourage proper configuration by warning if attributes aren't defined through ICustomAttribute
  const attributeNames = Object.keys(requiredAttributes);

  for (const attributeName of attributeNames) {
    const attribute = requiredAttributes[attributeName];

    // Verify that the attribute is properly configured
    if (!attribute) {
      throw new Error(`Required attribute ${attributeName} is not properly defined`);
    }

    // Additional validations could go here, but we're limited by CDK's public APIs
    // For now, we just ensure the attribute exists in the required attributes collection
  }
}
