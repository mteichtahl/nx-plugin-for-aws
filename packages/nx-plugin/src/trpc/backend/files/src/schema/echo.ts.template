import { z } from 'zod';

export const EchoInputSchema = z.object({
  message: z.string(),
});

export type IEchoInput = z.TypeOf<typeof EchoInputSchema>;

export const EchoOutputSchema = z.object({
  result: z.string(),
});

export type IEchoOutput = z.TypeOf<typeof EchoOutputSchema>;
