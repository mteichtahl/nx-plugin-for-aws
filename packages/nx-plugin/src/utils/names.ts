/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import camelCase from 'lodash.camelcase';
import deburr from 'lodash.deburr';

export const toClassName = (str?: string): string => {
  if (!str) {
    return str;
  }
  const words = str.replace(/[^a-zA-Z0-9]/g, ' ').split(/\s+/);
  return words
    .map((word, index) => {
      if (index === 0 && /^\d/.test(word)) {
        return '_' + word;
      }
      return word.charAt(0).toUpperCase() + word.slice(1);
    })
    .join('');
};

export const toKebabCase = (str?: string): string =>
  str?.split('/').map(kebabCase).join('/');

export const toSnakeCase = (str?: string): string =>
  str?.split('/').map(snakeCase).join('/');

export const upperFirst = (str: string): string =>
  str.charAt(0).toUpperCase() + str.slice(1);

export const pascalCase = (str: string): string => upperFirst(camelCase(str));

export const snakeCase = (str: string): string => {
  return (
    deburr(str)
      // Replace series of special characters with underscores
      .replace(/[^a-zA-Z0-9]+/g, '_')
      // Replace capital letters preceded by lowercase or numbers with underscore + lowercase
      .replace(/([a-z0-9])([A-Z])/g, '$1_$2')
      // Convert to lowercase
      .toLowerCase()
      // Remove leading/trailing underscores
      .replace(/^_+|_+$/g, '')
  );
};

export const kebabCase = (str: string): string => {
  return (
    deburr(str)
      // Replace series of special characters with hyphens
      .replace(/[^a-zA-Z0-9]+/g, '-')
      // Replace capital letters preceded by lowercase or numbers with hyphen + lowercase
      .replace(/([a-z0-9])([A-Z])/g, '$1-$2')
      // Convert to lowercase
      .toLowerCase()
      // Remove leading/trailing hyphens
      .replace(/^-+|-+$/g, '')
  );
};

// Convert a string to a dot notation string (eg. lambda_handler/my_handler.py -> lambda_handler.my_handler)
export const toDotNotation = (str: string): string =>
  str
    ?.replace(/^\/|\/$/g, '') // Remove leading/trailing slashes
    .replace(/\.[^/.]+$/g, '') // Remove any file extensions
    .split('/')
    .filter(Boolean) // Remove empty segments (in case of double slashes)
    .join('.');
