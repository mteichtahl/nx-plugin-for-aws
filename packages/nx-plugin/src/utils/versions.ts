/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
export const VERSIONS = {
  '@cdklabs/cdk-validator-cfnguard': '^0.0.60',
  '@aws-sdk/client-cognito-identity': '^3.775.0',
  '@aws-sdk/credential-providers': '^3.775.0',
  '@aws-sdk/credential-provider-cognito-identity': '^3.775.0',
  '@aws-lambda-powertools/logger': '^2.17.0',
  '@aws-lambda-powertools/metrics': '^2.17.0',
  '@aws-lambda-powertools/tracer': '^2.17.0',
  '@nxlv/python': '^21.0.0',
  '@nx/devkit': '~21.0.3',
  '@modelcontextprotocol/sdk': '^1.11.3',
  '@tanstack/react-router': '^1.121.16',
  '@tanstack/router-plugin': '^1.121.16',
  '@tanstack/router-generator': '^1.121.16',
  '@tanstack/virtual-file-routes': '^1.120.17',
  '@tanstack/router-utils': '^1.121.0',
  '@cloudscape-design/board-components': '^3.0.94',
  '@cloudscape-design/components': '^3.0.928',
  '@cloudscape-design/global-styles': '^1.0.38',
  '@tanstack/react-query': '^5.74.3',
  '@trpc/tanstack-react-query': '11.0.0',
  '@trpc/client': '11.0.0',
  '@trpc/server': '11.0.0',
  '@types/node': '^22.13.13',
  '@types/aws-lambda': '^8.10.148',
  '@types/cors': '^2.8.18',
  '@smithy/types': '^4.2.0',
  aws4fetch: '^1.0.20',
  'aws-cdk': '^2.1006.0',
  'aws-cdk-lib': '^2.200.0',
  'aws-xray-sdk-core': '^3.10.3',
  constructs: '^10.4.2',
  cors: '^2.8.5',
  esbuild: '^0.25.1',
  'eslint-plugin-prettier': '^5.2.5',
  'jsonc-eslint-parser': '^2.4.0',
  'oidc-client-ts': '^3.2.0',
  prettier: '^3.5.3',
  'react-oidc-context': '^3.2.0',
  'source-map-support': '^0.5.21',
  tsx: '4.20.1', // https://github.com/privatenumber/tsx/issues/727
  'vite-tsconfig-paths': '^5.1.4',
  zod: '^3.25.50',
} as const;

/**
 * Add versions to the given dependencies
 */
export const withVersions = (deps: (keyof typeof VERSIONS)[]) =>
  Object.fromEntries(deps.map((dep) => [dep, VERSIONS[dep]]));
