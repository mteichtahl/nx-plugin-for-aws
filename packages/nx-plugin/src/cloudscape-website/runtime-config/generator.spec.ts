/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { Tree } from '@nx/devkit';
import {
  RUNTIME_CONFIG_GENERATOR_INFO,
  runtimeConfigGenerator,
} from './generator';
import { RuntimeConfigGeneratorSchema } from './schema';
import { createTreeUsingTsSolutionSetup } from '../../utils/test';
import { sharedConstructsGenerator } from '../../utils/shared-constructs';
import { expectHasMetricTags } from '../../utils/metrics.spec';

describe('runtime-config generator', () => {
  let tree: Tree;

  const options: RuntimeConfigGeneratorSchema = {
    project: 'test-app',
  };

  beforeEach(() => {
    tree = createTreeUsingTsSolutionSetup();
    tree.write(
      'packages/test-app/project.json',
      JSON.stringify({
        name: 'test-app',
        sourceRoot: 'packages/test-app/src',
      }),
    );
  });

  it('should generate runtime config files', async () => {
    // Set up a basic React app structure
    tree.write(
      'packages/test-app/src/main.tsx',
      `import { RouterProvider } from '@tanstack/react-router';

        export function App() {
        
          const App = () => <RouterProvider router={router} />;

          return (
            <App/>
          );
        }`,
    );
    await runtimeConfigGenerator(tree, options);
    // Check if RuntimeConfig component was generated
    expect(
      tree.exists('packages/test-app/src/components/RuntimeConfig/index.tsx'),
    ).toBeTruthy();
    // Snapshot the generated RuntimeConfig component
    expect(
      tree
        .read('packages/test-app/src/components/RuntimeConfig/index.tsx')
        ?.toString(),
    ).toMatchSnapshot('runtime-config-component.tsx');
  });

  it('should modify main.tsx correctly', async () => {
    // Set up a basic React app structure
    tree.write(
      'packages/test-app/src/main.tsx',
      `import { RouterProvider } from '@tanstack/react-router';

        export function App() {
        
          const App = () => <RouterProvider router={router} />;

          return (
            <App/>
          );
        }`,
    );
    await runtimeConfigGenerator(tree, options);
    const mainTsxContent = tree
      .read('packages/test-app/src/main.tsx')
      ?.toString();
    expect(mainTsxContent).toMatchSnapshot('modified-main.tsx');
  });

  it('should skip generation if RuntimeConfig already exists', async () => {
    // Set up a basic React app structure
    tree.write(
      'packages/test-app/src/main.tsx',
      `import { RouterProvider } from '@tanstack/react-router';

        export function App() {
        
          const App = () => <RouterProvider router={router} />;

          return (
            <App/>
          );
        }`,
    );
    // First run to generate files
    await runtimeConfigGenerator(tree, options);
    // Get file content after first run
    const firstRunContent = tree
      .read('packages/test-app/src/components/RuntimeConfig/index.tsx')
      ?.toString();
    // Modify the file to simulate manual changes
    tree.write(
      'packages/test-app/src/components/RuntimeConfig/index.tsx',
      'modified content',
    );
    // Run generator again
    await runtimeConfigGenerator(tree, options);
    // Verify file wasn't overwritten
    expect(
      tree
        .read('packages/test-app/src/components/RuntimeConfig/index.tsx')
        ?.toString(),
    ).toBe('modified content');
    expect(
      tree
        .read('packages/test-app/src/components/RuntimeConfig/index.tsx')
        ?.toString(),
    ).not.toBe(firstRunContent);
  });

  it('should throw error if main.tsx does not exist', async () => {
    await expect(runtimeConfigGenerator(tree, options)).rejects.toThrow(
      'Can only run this generator on a project which contains packages/test-app/src/main.tsx',
    );
  });

  it('should throw error if App is not found', async () => {
    // Set up main.tsx without App
    tree.write(
      'packages/test-app/src/main.tsx',
      `export function App() {
        return <div>Test App</div>;
      }`,
    );
    await expect(runtimeConfigGenerator(tree, options)).rejects.toThrow(
      'Could not locate the App element in main.tsx',
    );
  });

  it('should generate shared constructs', async () => {
    // Set up a basic React app structure
    tree.write(
      'packages/test-app/src/main.tsx',
      `import { RouterProvider } from '@tanstack/react-router';

        export function App() {
        
          const App = () => <RouterProvider router={router} />;

          return (
            <App/>
          );
        }`,
    );
    await runtimeConfigGenerator(tree, options);
    // Check if shared constructs were generated
    expect(
      tree.exists('packages/common/constructs/src/core/index.ts'),
    ).toBeTruthy();
    expect(
      tree.exists('packages/common/constructs/src/core/runtime-config.ts'),
    ).toBeTruthy();
    // Snapshot the shared constructs files
    expect(
      tree.read('packages/common/constructs/src/core/index.ts')?.toString(),
    ).toMatchSnapshot('common/constructs-index.ts');
    expect(
      tree
        .read('packages/common/constructs/src/core/runtime-config.ts')
        ?.toString(),
    ).toMatchSnapshot('runtime-config.ts');
  });

  it('should add generator metric to app.ts', async () => {
    // Set up test tree with shared constructs
    await sharedConstructsGenerator(tree);

    // Set up a basic React app structure
    tree.write(
      'packages/test-app/src/main.tsx',
      `import { RouterProvider } from '@tanstack/react-router';

        export function App() {
        
          const App = () => <RouterProvider router={router} />;

          return (
            <App/>
          );
        }`,
    );

    // Call the generator function
    await runtimeConfigGenerator(tree, options);

    // Verify the metric was added to app.ts
    expectHasMetricTags(tree, RUNTIME_CONFIG_GENERATOR_INFO.metric);
  });
});
