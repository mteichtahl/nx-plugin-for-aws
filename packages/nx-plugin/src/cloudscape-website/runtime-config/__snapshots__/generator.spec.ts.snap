// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`runtime-config generator > should generate runtime config files > runtime-config-component.tsx 1`] = `
"import { Spinner } from '@cloudscape-design/components';
import React, {
  createContext,
  PropsWithChildren,
  useEffect,
  useState,
} from 'react';
import type { IRuntimeConfig } from ':proj/common-types';

/**
 * Context for storing the runtimeConfig.
 */
export const RuntimeConfigContext = createContext<IRuntimeConfig | undefined>(
  undefined,
);

/**
 * Apply any overrides to point to local servers/resources here
 * for the serve-local target
 */
const applyOverrides = (runtimeConfig: IRuntimeConfig) => {
  if (import.meta.env.MODE === 'serve-local') {
    // Add local server urls here
  }
  return runtimeConfig;
};

/**
 * Sets up the runtimeConfig.
 *
 * This assumes a runtime-config.json file is present at '/'.
 */
const RuntimeConfigProvider: React.FC<PropsWithChildren> = ({ children }) => {
  const [runtimeConfig, setRuntimeConfig] = useState<
    IRuntimeConfig | undefined
  >();
  useEffect(() => {
    (async () => {
      try {
        setRuntimeConfig(
          applyOverrides(await (await fetch('/runtime-config.json')).json()),
        );
      } catch {
        setRuntimeConfig(applyOverrides({ apis: {} } as any));
      }
    })();
  }, [setRuntimeConfig]);

  return runtimeConfig ? (
    <RuntimeConfigContext.Provider value={runtimeConfig}>
      {children}
    </RuntimeConfigContext.Provider>
  ) : (
    <Spinner />
  );
};

export default RuntimeConfigProvider;
"
`;

exports[`runtime-config generator > should generate shared constructs > common/constructs-index.ts 1`] = `
"export * from './app.js';
export * from './runtime-config.js';
"
`;

exports[`runtime-config generator > should generate shared constructs > runtime-config.ts 1`] = `
"import type { IRuntimeConfig } from ':proj/common-types';
import { Stack } from 'aws-cdk-lib';
import { Construct } from 'constructs';

const RuntimeConfigKey = '__RuntimeConfig__';

export class RuntimeConfig extends Construct {
  private readonly _runtimeConfig: Partial<IRuntimeConfig> = {};

  static ensure(scope: Construct): RuntimeConfig {
    const stack = Stack.of(scope);
    return (
      RuntimeConfig.of(scope) ?? new RuntimeConfig(stack, RuntimeConfigKey)
    );
  }

  static of(scope: Construct): RuntimeConfig | undefined {
    const stack = Stack.of(scope);
    return stack.node.tryFindChild(RuntimeConfigKey) as
      | RuntimeConfig
      | undefined;
  }

  constructor(scope: Construct, id: string) {
    super(scope, id);
  }

  get config(): Partial<IRuntimeConfig> {
    return this._runtimeConfig;
  }
}
"
`;

exports[`runtime-config generator > should modify main.tsx correctly > modified-main.tsx 1`] = `
"import RuntimeConfigProvider from './components/RuntimeConfig';
import { RouterProvider } from '@tanstack/react-router';

export function App() {
  const App = () => <RouterProvider router={router} />;

  return (
    <RuntimeConfigProvider>
      <App />
    </RuntimeConfigProvider>
  );
}
"
`;
